File: 0103-binary-tree-zigzag-level-order-traversal/0103-binary-tree-zigzag-level-order-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        List<List<Integer>> ans = new ArrayList<>();
        if (root == null) return ans;

        Queue<TreeNode> temp = new LinkedList<>();
        temp.offer(root);
        boolean ltor = true;

        while (!temp.isEmpty()) {
            int l = temp.size();
            List<Integer> tree = new ArrayList<>(Collections.nCopies(l, 0));

            for (int i = 0; i < l; i++) {
                TreeNode node = temp.poll();
                tree.set(ltor ? i : l - 1 - i, node.val);

                if (node.left != null) temp.offer(node.left);
                if (node.right != null) temp.offer(node.right);
            }

            ltor = !ltor;
            ans.add(tree);
        }

        return ans;
        
    }
}

File: LeetCode/Medium/0103-binary-tree-zigzag-level-order-traversal/0103-binary-tree-zigzag-level-order-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def zigzagLevelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        if not root: return []
        dq = deque()
        reverse = False
        res = []
        dq.append(root)
        while dq:
            level = []
            for _ in range(len(dq)):
                if not reverse:
                    temp = dq.popleft()
                    level.append(temp.val)
                    if temp.left:
                        dq.append(temp.left)
                    if temp.right:
                        dq.append(temp.right)
                else:
                    temp = dq.pop()
                    level.append(temp.val)
                    if temp.right:
                        dq.appendleft(temp.right)
                    if temp.left:
                        dq.appendleft(temp.left)
            reverse = not reverse
            res.append(level)
        return res

